### PR TITLE
Feat: E2E Docker Support

### PR DESCRIPTION
**Pull request by seriouslysean**
## Issue number

Relevant [issue number](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- Resolves #

## Please check the following

- [ ] Do the tests still pass? (see [Run the Tests](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#run-the-tests))
- [ ] Is the code formatted properly? (see [Linting (Formatting)](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#linting-formatting))
- For New Features:
  - [ ] Have tests been added to cover any new features or fixes?
  - [ ] Has the documentation been updated accordingly?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,6 +68,7 @@
     "docker:nuke": "docker rmi $(docker images -f=\"reference=cuttle*\" -q --no-trunc) || true; npm run docker:clean",
     "docker:start": "docker compose -f ./docker/docker-compose.yml up",
     "docker:stop": "docker compose -f ./docker/docker-compose.yml down",
+    "docker:e2e": "docker compose -f ./docker/docker-compose.yml -f ./docker/docker-compose-e2e.yml up --exit-code-from e2e",
```
Comment (seriouslysean) [inline package.json:71]: @itsalaidbacklife you can run this to try it out but the tests all fail:
```
e2e       | http://localhost:1337/test/wipeDatabase
e2e       | 
e2e       | We attempted to make an http request to this URL but the request failed without a response.
e2e       | 
e2e       | We received this error at the network level:
e2e       | 
e2e       |   > Error: connect ECONNREFUSED 127.0.0.1:1337
e2e       | 
e2e       | -----------------------------------------------------------
e2e       | 
e2e       | The request we sent was:
e2e       | 
e2e       | Method: GET
e2e       | URL: http://localhost:1337/test/wipeDatabase
e2e       | 
e2e       | -----------------------------------------------------------
e2e       | 
e2e       | Common situations why this would fail:
e2e       |   - you don't have internet access
e2e       |   - you forgot to run / boot your web server
e2e       |   - your web server isn't accessible
e2e       |   - you have weird network configuration settings on your computer
e2e       | 
e2e       | https://on.cypress.io/request
e2e       | 
e2e       | Because this error occurred during a `before each` hook we are skipping the remaining tests in the current suite: `Signing Up`
e2e       |       at <unknown> (http://client:8080/__cypress/runner/cypress_runner.js:149355:78)
e2e       |       at tryCatcher (http://client:8080/__cypress/runner/cypress_runner.js:18744:23)
e2e       |       at <unknown> (http://client:8080/__cypress/runner/cypress_runner.js:13883:37)
e2e       |       at tryCatcher (http://client:8080/__cypress/runner/cypress_runner.js:18744:23)
e2e       |       at Promise._settlePromiseFromHandler (http://client:8080/__cypress/runner/cypress_runner.js:16679:31)
e2e       |       at Promise._settlePromise (http://client:8080/__cypress/runner/cypress_runner.js:16736:18)
e2e       |       at Promise._settlePromise0 (http://client:8080/__cypress/runner/cypress_runner.js:16781:10)
e2e       |       at Promise._settlePromises (http://client:8080/__cypress/runner/cypress_runner.js:16857:18)
e2e       |       at _drainQueueStep (http://client:8080/__cypress/runner/cypress_runner.js:13451:12)
e2e       |       at _drainQueue (http://client:8080/__cypress/runner/cypress_runner.js:13444:9)
e2e       |       at ../../node_modules/bluebird/js/release/async.js.Async._drainQueues (http://client:8080/__cypress/runner/cypress_runner.js:13460:5)
e2e       |       at Async.drainQueues (http://client:8080/__cypress/runner/cypress_runner.js:13330:14)
e2e       |   From Your Spec Code:
e2e       |       at Context.eval (webpack:///./tests/e2e/support/commands.js:15:5)
e2e       |   
e2e       |   From Node.js Internals:
e2e       |     RequestError: Error: connect ECONNREFUSED 127.0.0.1:1337
e2e       |         at new RequestError (/root/.cache/Cypress/12.17.3/Cypress/resources/app/node_modules/request-promise-core/lib/errors.js:14:15)
e2e       |         at Request.plumbing.callback (/root/.cache/Cypress/12.17.3/Cypress/resources/app/node_modules/request-promise-core/lib/plumbing.js:87:29)
e2e       |         at Request.RP$callback [as _callback] (/root/.cache/Cypress/12.17.3/Cypress/resources/app/node_modules/request-promise-core/lib/plumbing.js:46:31)
e2e       |         at self.callback (/root/.cache/Cypress/12.17.3/Cypress/resources/app/node_modules/@cypress/request/request.js:183:22)
e2e       |         at Request.emit (node:events:527:28)
e2e       |         at Request.emit (node:domain:475:12)
e2e       |         at Request.onRequestError (/root/.cache/Cypress/12.17.3/Cypress/resources/app/node_modules/@cypress/request/request.js:869:8)
e2e       |         at ClientRequest.emit (node:events:539:35)
e2e       |         at ClientRequest.emit (node:domain:475:12)
e2e       |         at Socket.socketErrorListener (node:_http_client:454:9)
e2e       |         at Socket.emit (node:events:527:28)
e2e       |         at Socket.emit (node:domain:475:12)
e2e       |         at emitErrorNT (node:internal/streams/destroy:157:8)
e2e       |         at emitErrorCloseNT (node:internal/streams/destroy:122:3)
e2e       |         at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -50,12 +50,15 @@ services:
     command: npm run start:server
     ports:
       - "1337:1337"
+    working_dir: /cuttle
```
Comment (seriouslysean) [inline docker/docker-compose.yml:53]: Moved working dir to the compose file.

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -50,12 +50,15 @@ services:
...
     volumes: *app-volume
     depends_on:
       - database
 
   client:
     container_name: client
+    # Required to be able to hit http://client:8080 from the Cypress docker container
+    hostname: client
```
Comment (seriouslysean) [inline docker/docker-compose.yml:61]: I'm not sold on this yet, it's nice to have for image<>image communication, so probably worth leaving. May also just add it to the e2e override file.

File: docker/docker-compose-e2e.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# This file is only meant to be run as an override to docker-compose.yml
```
Comment (seriouslysean) [inline docker/docker-compose-e2e.yml:1]: This is really interesting-- you can extend compose files with others and they'll smartly merge. So we get the benefit of all 3 normal services (db, server, client) and can add an e2e service, also.

File: docker/Cypress.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# Sync this up with the Cypress version in package.json
```
Comment (seriouslysean) [inline docker/Cypress.Dockerfile:1]: I tried to use the Cypress official image with included support and I don't know how to get it to work. May need to let it marinate a bit. For now I'm installing all the required deps from their docker image recommendation. `npm ci` fails in theirs due to a missing Python dep. 🤷🏼

**Review by seriouslysean — COMMENTED** [review]
Far from finished, but it's a start.

File: docker/docker-compose-e2e.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+# see https://docs.docker.com/compose/extends/#multiple-compose-files
+services:
+  client:
+    environment:
+      - CUTTLE_DOCKERIZED=true
+      - VITE_API_URL=http://server:1337
+      - VITE_BUILD_OUTPUT=assets
```
Comment (itsalaidbacklife) [inline docker/docker-compose-e2e.yml:8]: If this file gets merged with the existing `docker-compose.yml` why does it need to define client here?

File: docker/docker-compose.yml (hunk 3 above)
Comment (itsalaidbacklife) [inline docker/docker-compose.yml:61]: Why is `hostname` necessary for the client DNS but not for database, which is also used for container to container communication?

**Review by itsalaidbacklife — COMMENTED** [review]
Nice start. My first suspicion about the failures is that the `io` instanve in the cypress commands file is hard coded to connect to localhost:1337 but when running in the e2e docker container it instead needs to connect to http://server:1337 just like the client service does.

We could likely fix this by defining the test code's server url as an env var and having the test code consume it in commands.js and fall back to localhost:1337 when unspecified

File: docker/docker-compose-e2e.yml (hunk 6 above)
Comment (seriouslysean) [reply docker/docker-compose-e2e.yml:8]: It's not merged, it overrides the key, so in this case we need all the same values as far as I understand. I could be wrong though, the documentation is a little hard to parse. I can probably check it to be honest, I'll make a note to verify.

File: docker/docker-compose.yml (hunk 3 above)
Comment (seriouslysean) [reply docker/docker-compose.yml:61]: This is tricky to figure out-- it seems like if the app or service is running INSIDE of docker, it can use these hostnames to talk amongst the other shared clients. When you expose to the host (your local) you can only use the ports and none of the internal hostnames.

**Comment by itsalaidbacklife** [comment]
Closing as this isn't actively worked on
